command/certificate/handshake.go

```diff
@@ -0,0 +1,260 @@
+		MinVersion:         minVersion,
+		MaxVersion:         maxVersion,
+		RootCAs:            rootCAs,
+		InsecureSkipVerify: insecure,
+		ServerName:         serverName,
+		Certificates:       certificates,
+	}
+
+	cs, err := tlsDialWithFallback(addr, tlsConfig)
+	if err != nil {
+		return err
+	}
+
+	// Print only the list of verified chains
+	if printChains {
+		for _, chain := range cs.VerifiedChains {
+			for _, crt := range chain {
+				fmt.Print(string(pem.EncodeToMemory(&pem.Block{
+					Type:  "CERTIFICATE",
+					Bytes: crt.Raw,
+				})))
+			}
+		}
+		return nil
+	}
+
+	// Print only the peer certificates
+	if printPeer {
+		for _, crt := range cs.PeerCertificates {
+			fmt.Print(string(pem.EncodeToMemory(&pem.Block{
+				Type: "CERTIFICATE", Bytes: crt.Raw,
+			})))
+		}
+		return nil
+	}
+
+	// Check if the certificates is verified
+	var intermediates *x509.CertPool
+	if len(cs.PeerCertificates) > 1 {
+		intermediates = x509.NewCertPool()
+		for _, crt := range cs.PeerCertificates[1:] {
+			intermediates.AddCert(crt)
+		}
+	}
+	_, verifyErr := cs.PeerCertificates[0].Verify(x509.VerifyOptions{
+		Roots:         rootCAs,
+		Intermediates: intermediates,
+		DNSName:       serverName,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+	})
+
+	connStateValue := reflect.ValueOf(cs)
+	curveIDField := connStateValue.FieldByName("testingOnlyCurveID")
+
+	fmt.Printf("Server Name:  %s\n", cs.ServerName)
+	fmt.Printf("Version:      %s\n", tls.VersionName(cs.Version))
+	fmt.Printf("Cipher Suite: %s\n", tls.CipherSuiteName(cs.CipherSuite))
+	fmt.Printf("KEM:          %s\n", curveIDName(curveIDField.Uint()))
+	fmt.Printf("Insecure:     %v\n", tlsConfig.InsecureSkipVerify)
+	fmt.Printf("Verified:     %v\n", verifyErr == nil)
+
+	return nil
+}
+
+func curveIDName(curveID uint64) string {
+	switch tls.CurveID(curveID) {
+	case tls.CurveP256:
+		return "P-256"
+	case tls.CurveP384:
+		return "P-384"
+	case tls.CurveP521:
+		return "P-521"
+	case tls.X25519:
+		return "X25519"
+	case tls.X25519MLKEM768:
+		return "X25519MLKEM768"
+	default:
+		return "Unknown"
+	}
+}
+
+func getTLSVersions(s string) (uint16, uint16, error) {
+	switch s {
+	case "":
+		return tls.VersionTLS12, 0, nil
+	case "1.3":
+		return tls.VersionTLS13, tls.VersionTLS13, nil
+	case "1.2":
+		return tls.VersionTLS12, tls.VersionTLS12, nil
+	case "1.1":
+		return tls.VersionTLS11, tls.VersionTLS11, nil
+	case "1.0":
+		return tls.VersionTLS10, tls.VersionTLS10, nil
+	default:
+		return 0, 0, fmt.Errorf("unsupported TLS version %q", s)
+	}
+}
+
+func tlsDialWithFallback(addr string, tlsConfig *tls.Config) (tls.ConnectionState, error) {
+	conn, err := tls.Dial("tcp", addr, tlsConfig)
+	if err != nil {
+		if tlsConfig.InsecureSkipVerify {
+			return tls.ConnectionState{}, fmt.Errorf("error connecting to %q: %w", addr, err)
+		}
+		tlsConfig.InsecureSkipVerify = true
+		return tlsDialWithFallback(addr, tlsConfig)
+	}
+	defer conn.Close()
+	conn.Handshake()
```

Comment:
Can use `HandshakeContext`, and it would be good to check and return the error. 

In a follow up we could implement some additional error handling logic for more informative errors based on some internal code we have. I think it could be nice to put that in `tlsutil`.

Reply:
Fixed with 6a13fa36c1ed44304247d9a7c64b7f0495fae9e6